src/java.desktop/share/classes/javax/swing/plaf/basic/BasicDirectoryModel.java

```diff
@@ -522,45 +515,46 @@ public void run() {
     private final class DoChangeContents implements Runnable {
         private final List<File> addFiles;
         private final List<File> remFiles;
-        private boolean doFire = true;
         private final int fid;
-        private int addStart = 0;
-        private int remStart = 0;
+        private final int addStart;
+        private final int remStart;
 
-        DoChangeContents(List<File> addFiles, int addStart, List<File> remFiles,
-                         int remStart, int fid) {
+        private DoChangeContents(List<File> addFiles, int addStart,
+                                 List<File> remFiles, int remStart,
+                                 int fid) {
             this.addFiles = addFiles;
             this.addStart = addStart;
             this.remFiles = remFiles;
             this.remStart = remStart;
             this.fid = fid;
         }
 
-        synchronized void cancel() {
-            doFire = false;
-        }
+        @Override
+        public void run() {
+            if (fetchID.get() != fid) {
```

Comment:
How it will work if the "FilesLoader" was created already but before loading the files the "invalidateFileCache" was called, before the patch we stop execution due to "doFire=false".

Reply:
Good catch!

I'll bump `fetchID` in `invalidateFileCache`. I missed this part.